### ISSUE TITLE
Align VI/MFVI sample reuse with ABRIS batch-sequential algorithm

### ISSUE THREAD

**Issue by eparish1** [Pressio/rom-tools-and-workflows#395]
## Summary

Bring the current VI/MFVI importance-sampling sample-reuse implementation closer to the **adaptive batch-sequential regulated importance sampling (ABRIS)** algorithm described by Robalo Rei et al. (this is the method currently referred to as `ARBIS` in our benchmark discussions).

The current implementation in `romtools/workflows/inverse/vi_sample_reuse.py` is intentionally a first-pass ABRIS-style implementation: it keeps a moving archive, uses deterministic-mixture importance weights, applies ESS / score / periodic refresh criteria, and supports VI and coupled MFVI. However, it currently makes a **single reuse-vs-refresh decision per optimizer iteration**. The published ABRIS algorithm instead uses an **inner batch-sequential sampling loop**: if the retained sample set is inadequate, draw one fresh batch of size `N`, update the retained sets / mixture proposal, recompute the diagnostics, and repeat until the sample set passes or the moving-window/batch limit is reached.

This issue should make the single-fidelity implementation follow that algorithm more faithfully and define a mathematically consistent extension for MFVI rather than simply applying the single-fidelity logic around the existing MF estimator.

## Current implementation

`vi_sample_reuse.py` currently provides:

- `VISampleReuseConfig` with `history_batches`, ESS threshold, score diagnostic, and periodic refresh;
- a moving `_ReuseArchive` of FOM evaluations and the variational distribution that generated each batch;
- deterministic-mixture importance weights with component weights proportional to retained batch size;
- an origin-proposal-weighted LOO baseline;
- one refresh/reuse decision per call to `_evaluate_vi_state` / `_evaluate_mf_vi_state`;
- for MFVI, reuse of the FOM archive while reevaluating the current ROM on the archived FOM parameter points and drawing fresh ROM-only samples from the current variational distribution.

The file itself notes that this is a first-pass implementation and that a new multifidelity gain derivation was deferred.

## Desired single-fidelity behavior

Refactor the reuse controller to follow the ABRIS sampling loop more directly.

At optimizer iteration `i`:

1. Construct the retained moving window from batches generated by recent variational distributions.
2. Form the deterministic-mixture proposal

   `q_IS^i(theta) = sum_j beta_j^i q(theta | lambda_j)`

   with

   `beta_j^i = n_j / sum_k n_k`.

3. Compute importance weights for the current `q(theta | lambda_i)`.
4. Evaluate the ABRIS adequacy criteria on the current retained set:
   - ESS criterion;
   - score-function diagnostic;
   - mandatory periodic sampling / exploration criterion.
5. If the criteria are not satisfied, draw **one new batch of exactly `N` samples from the current variational distribution**, evaluate the expensive model, append the batch, rebuild the proposal, and re-evaluate the criteria.
6. Continue the inner batch-sequential loop until:
   - the sample set passes the criteria, or
   - the configured maximum moving-window / batch count is reached.
7. If the limit is reached, the current iteration should naturally reduce to a large fresh/current-distribution BBVI estimate rather than accepting an inadequate historical archive.
8. Use the accepted retained sample set for the gradient / ELBO estimate and then perform the optimizer update.

The implementation should preserve batch-sized model execution so `N` can be chosen to match available parallel resources.

### Moving-window semantics

Match the paper more closely:

- `m` should represent the maximum number of retained sampling batches/components (or expose an equivalently clear parameter);
- early iterations should use `min(m, available iterations/batches)` naturally;
- old components should be removed as the moving window advances;
- newly drawn batches at the current iteration should remain distinguishable so multiple inner-loop batches from one optimizer iteration are handled correctly;
- mixture coefficients should be recomputed from actual retained batch sizes.

Consider renaming or documenting `history_batches` so its semantics are unambiguous relative to the ABRIS moving-window parameter `m`.

## Multifidelity extension: design carefully

The MFVI case is the main technical concern in this issue.

The existing MF estimator assumes a coupled high-/low-fidelity base set plus additional low-fidelity samples. Without sample reuse it has the form

`E_hat_H + A (E_hat_L,full - E_hat_L,base)`.

With historical FOM reuse, the base set is no longer an i.i.d. sample from the current variational distribution: it is a heterogeneous mixture of samples from previous `q_j` distributions and must be importance weighted. The current code does this for the HF and coupled LF terms, then combines those estimates with fresh ROM-only samples using the existing hierarchical MF correction. This is plausible, but it is **not automatically the same estimator/optimal gain derivation as standard MFMC**, and the implementation currently acknowledges that the new MF gain derivation was deferred.

Before treating `ARBIS + MF` as a production algorithm, explicitly derive and document the estimator being used.

### Recommended MF interpretation

Treat ABRIS primarily as the **HF sampling/reuse policy** and construct the MF correction around estimators that all target expectations under the current variational distribution `q_i`:

- Historical paired HF samples:
  - retain FOM outputs permanently;
  - evaluate the **current** LF/ROM model on the retained FOM parameter points when needed (cheap relative to FOM);
  - use the same deterministic-mixture importance weights for both HF and coupled-LF score terms.

- Additional LF samples:
  - draw from the current `q_i` (initial implementation), or introduce a separately derived LF reuse archive later;
  - keep this estimator logically separate from the historical HF archive.

Then derive the control-variate estimator from these two estimators rather than assuming the original i.i.d. nested-sample MFMC formula still applies unchanged.

### MF questions that must be resolved

1. **Control-variate gain (`alpha`)**
   - Derive the variance-minimizing coefficient for the actual estimator with MIS-weighted historical HF/LF pairs plus current-distribution LF-only samples.
   - Do not assume `_compute_mfmc_alpha(weighted_high_terms, weighted_low_terms)` remains optimal after reuse.
   - Decide whether the initial production implementation uses an estimated gain or a fixed/safeguarded gain.

2. **Nested-sample scaling / current `ratio` factor**
   - The current reuse path applies `n_extra / (n_archive + n_extra)` to the LF correction, mirroring the ordinary nested MFMC estimator.
   - Verify this is the intended variance-reduction estimator once the base mean is an importance-sampling estimate from heterogeneous proposals rather than an ordinary sample mean from `q_i`.
   - If the correction is instead formed as the difference of two unbiased estimates of the same current LF expectation, derive the appropriate scaling and gain together.

3. **Bias from an estimated gain**
   - An `alpha` estimated from the same samples used in the control-variate estimate is generally not exactly unbiased.
   - Quantify/acknowledge this, or estimate the gain using lagged / independent information if exact unbiasedness is a requirement.
   - This is separate from the importance-sampling correctness of the HF estimate.

4. **LOO baseline with reused samples**
   - Verify analytically that the current origin-weighted LOO construction has the desired expectation when samples come from multiple proposal components.
   - Ensure the same baseline convention is used consistently for the HF and coupled LF score terms.
   - Fresh LF-only samples can use the ordinary current-distribution LOO baseline.

5. **ROM rebuilds / changing LF model**
   - Historical FOM evaluations remain valid forever, but historical LF evaluations do not if the auto-ROM changes.
   - Do **not** cache stale LF values across ROM rebuilds unless model-version provenance is explicitly tracked.
   - Prefer reevaluating the current ROM on retained HF parameter samples after a rebuild.

6. **Reuse acceptance criterion in MFVI**
   - The ABRIS ESS and score criteria diagnose adequacy of the importance-sampled HF/current-target estimator.
   - Decide whether MFVI also needs an additional criterion based on HF/LF correlation, control-variate effectiveness, or ROM error.
   - Poor LF correlation should disable/reduce the MF correction, but should not by itself invalidate a statistically adequate HF ABRIS archive.

7. **Adaptive ROM training data**
   - Only genuinely new FOM evaluations should enlarge the ROM training set.
   - Reused historical FOM samples must not be double-counted as new training observations.
   - FOM accounting should distinguish fresh ABRIS batches from reused evaluations.

8. **Multiple fresh HF batches in one optimizer iteration**
   - If the ABRIS inner loop needs 2+ fresh batches at the current `lambda_i`, all of those FOM samples should be available to the adaptive ROM/training-data update.
   - Ensure directory/provenance/restart handling supports multiple batches at one optimizer iteration.

### In scope

- Proper inner ABRIS batch-sequential sampling loop for gradient-based VI.
- Moving-window/component bookkeeping matching the published algorithm.
- ESS, score and periodic criteria evaluated after every added batch.
- Correct diagnostics and model-call accounting for zero, one, or multiple fresh batches per optimizer iteration.
- MFVI extension with a documented/derived MIS + control-variate estimator.
- Auto-ROM compatibility, including ROM rebuilds and FOM training-data provenance.
- LOO baseline correctness for heterogeneous reused samples.
- Unit tests and small statistical regression tests.

### Out of scope for this issue unless needed by the derivation

- RQMC reuse.
- Newton/Hessian reuse.
- Reusing historical LF-only samples from their own proposal archive.
- Major changes to the existing VI optimizer APIs.

## API / configuration

Evolve `VISampleReuseConfig` (name can remain for compatibility) to expose the ABRIS concepts directly. For example:

```python
VISampleReuseConfig(
    enabled=True,
    batch_size=None,               # default to VI/MFVI HF sample size
    max_window_batches=10,         # ABRIS m
    ess_threshold=None,
    use_score_diagnostic=True,
    score_error_scale=1.0,
    periodic_refresh=100,
    max_batches_per_iteration=None, # default consistent with m
)
```

Maintain backward compatibility for `history_batches` if it is renamed, with a deprecation warning.

The normal VI/MFVI sample-size argument should remain the natural default batch size `N`; do not introduce two competing sample-size settings unless there is a clear need.

## Diagnostics

Persist enough information to reproduce the ABRIS behavior:

- optimizer iteration;
- number of retained mixture components/batches;
- number of retained samples;
- number of new FOM batches added this iteration;
- number of new FOM evaluations added this iteration;
- cumulative FOM evaluations;
- ESS after each inner-loop check;
- score diagnostic after each inner-loop check;
- refresh/add-batch reason (`empty`, `ess`, `score`, `periodic`, etc.);
- mixture weights `beta_j` or sufficient batch metadata to reconstruct them;
- MFVI: current alpha/control-variate diagnostics, HF/LF correlation diagnostic, ROM error/rebuild status.

### VI

- No-refresh iteration performs zero new FOM evaluations.
- Failed ESS criterion adds exactly one `N`-sample batch and re-tests.
- Repeated failures can add multiple batches in one optimizer iteration.
- Sampling stops once criteria pass.
- Moving-window truncation gives the expected mixture components and `beta_j` values.
- Periodic criterion forces a fresh batch even when ESS/score pass.
- MIS estimate agrees statistically with direct sampling from the current `q` on an analytic problem.
- LOO reuse estimator is tested against a brute-force/reference expectation.

### MFVI

- With control variate disabled, MFVI + reuse reduces to the single-fidelity ABRIS HF estimator (apart from harmless LF diagnostics/evaluations).
- With LF == HF, the derived MF estimator behaves as predicted and strongly reduces variance.
- With an uncorrelated LF model, the correction tends toward zero / does not materially increase variance after safeguards.
- Rebuilding the ROM causes retained HF points to be reevaluated with the new ROM rather than using stale LF values.
- Reused FOM samples are not double-counted as new auto-ROM training data.
- Multiple fresh ABRIS batches in one optimizer iteration are all incorporated exactly once into training data.
- Numerical tests compare empirical estimator mean against a high-sample reference to expose unintended bias.

## Acceptance criteria

- [ ] VI uses an ABRIS-style inner batch-sequential sampling loop rather than one binary refresh decision per optimizer iteration.
- [ ] Fresh FOM calls are always made in batches of `N` and diagnostics are recomputed after each batch.
- [ ] The retained proposal is a moving mixture of the appropriate historical/current variational distributions with batch-size-proportional coefficients.
- [ ] Periodic exploration behavior matches the intended ABRIS semantics.
- [ ] VI diagnostics expose every inner-loop add-batch decision and FOM cost.
- [ ] MFVI has a written derivation for the MIS-weighted HF + LF control-variate estimator used in code.
- [ ] The MF control-variate gain and any nested-sample scaling are derived for that estimator rather than inherited without justification from ordinary MFMC.
- [ ] The expected-value/bias implications of estimating `alpha` from the same samples are explicitly documented and tested.
- [ ] Historical HF points are reevaluated with the current ROM after ROM changes; stale LF values are not silently reused.
- [ ] Reused HF samples are not counted as new adaptive-ROM training data.
- [ ] Statistical tests demonstrate agreement with a reference estimator on simple analytic problems.
- [ ] Existing non-reuse VI/MFVI behavior is unchanged.
- [ ] Existing public sample-reuse configuration remains backward compatible or has a clear deprecation path.

## Reference

Robalo Rei, Schmidt, Nitzler, Dinkel, and Wall, *A black box variational inference scheme for inverse problems with demanding physics-based models*, Advanced Modeling and Simulation in Engineering Sciences (2026). The relevant sections are the importance-sampling-enhanced BBVI formulation and the adaptive batch-sequential regulated importance sampling (ABRIS) sample-set update algorithm.
